client/client-multi/client-c/src/platform-impl/kaa_tcp_channel.c

```diff
@@ -0,0 +1,1349 @@
+    return error_code;
+}
+
+
+kaa_error_t kaa_tcp_channel_get_max_timeout(kaa_transport_channel_interface_t *self, uint16_t *max_timeout)
+{
+    KAA_RETURN_IF_NIL2(self, self->context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *tcp_channel = (kaa_tcp_channel_t *) self->context;
+
+    *max_timeout = tcp_channel->keepalive.keepalive_interval / 2;
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_tcp_channel_check_keepalive(kaa_transport_channel_interface_t *self)
+{
+    KAA_RETURN_IF_NIL2(self, self->context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *tcp_channel = (kaa_tcp_channel_t *) self->context;
+
+    kaa_error_t error_code = KAA_ERR_NONE;
+
+    KAA_LOG_INFO(tcp_channel->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) checking keepalive", tcp_channel->access_point.id);
+
+    if (tcp_channel->keepalive.keepalive_interval == 0
+            || tcp_channel->channel_state != KAA_TCP_CHANNEL_AUTHORIZED) {
+        return error_code;
+    }
+
+    kaa_time_t interval = (kaa_time_t) ext_get_systime() - (kaa_time_t) tcp_channel->keepalive.last_sent_keepalive;
+
+    if (interval >= (tcp_channel->keepalive.keepalive_interval / 2)) {
+        //Send ping request
+        error_code = kaa_tcp_channel_ping(tcp_channel);
+    }
+
+    return error_code;
+}
+
+/*
+ * Set socket events callbacks.
+ */
+kaa_error_t kaa_tcp_channel_set_socket_events_callback(kaa_transport_channel_interface_t *self, on_kaa_tcp_channel_event_fn callback, void *context)
+{
+    KAA_RETURN_IF_NIL2(self, self->context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *tcp_channel = (kaa_tcp_channel_t *) self->context;
+
+    tcp_channel->event_callback = callback;
+    tcp_channel->event_context = context;
+
+    KAA_LOG_INFO(tcp_channel->logger, KAA_ERR_NONE, "Kaa TCP channel(%d) set socket events callbacks", tcp_channel->access_point.id);
+
+    return KAA_ERR_NONE;
+
+}
+
+kaa_error_t kaa_tcp_channel_set_keepalive_timeout(kaa_transport_channel_interface_t *self, uint16_t keepalive)
+{
+    KAA_RETURN_IF_NIL2(self, self->context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t *tcp_channel = (kaa_tcp_channel_t *)self->context;
+
+    tcp_channel->keepalive.keepalive_interval = keepalive;
+
+    KAA_LOG_INFO(tcp_channel->logger,KAA_ERR_NONE,"Kaa TCP channel(%d) keepalive is set to %d seconds", tcp_channel->access_point.id, tcp_channel->keepalive.keepalive_interval);
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_tcp_channel_disconnect(kaa_transport_channel_interface_t  *self)
+{
+    KAA_RETURN_IF_NIL2(self, self->context, KAA_ERR_BADPARAM);
+    kaa_tcp_channel_t * tcp_channel = (kaa_tcp_channel_t *) self->context;
+
+    return kaa_tcp_channel_disconnect_internal(tcp_channel, KAATCP_DISCONNECT_NONE);
+}
+
+void kaa_tcp_channel_connack_message_callback(void *context, kaatcp_connack_t message)
+{
+    kaa_tcp_channel_t *channel = (kaa_tcp_channel_t *) context;
+    KAA_RETURN_IF_NIL(channel,);
```

Comment:
I think the 'null check' macros should be used the first.
This comment is for all Kaa TCP callback-s.